network/test/middleware_test.go

```diff
@@ -26,11 +28,31 @@ import (
 
 const testChannel = "test-channel"
 
+type TagsObserver struct {
+	tags chan string
+}
+
+func (co *TagsObserver) OnNext(peertag interface{}) {
+	pt, ok := peertag.(PeerTag)
+
+	if ok {
+		co.tags <- fmt.Sprintf("peer: %v tag: %v", pt.peer, pt.tag)
+	}
+
+}
+func (co *TagsObserver) OnError(err error) {
+	close(co.tags)
+}
+func (co *TagsObserver) OnComplete() {
+	close(co.tags)
+}
```

Comment:
Better to differentiate graceful vs un-graceful termination by logging the error at the error level